org/mozilla/jss/pkcs11/PK11Signature.c

```diff
@@ -48,39 +55,79 @@ Java_org_mozilla_jss_pkcs11_PK11Signature_initSigContext
     SGNContext *ctxt=NULL;
     jobject contextProxy=NULL;
     SECKEYPrivateKey *privk;
+    SECAlgorithmID *signAlg=NULL;
+    SECStatus rv;
+    int keyTypeOKForPSS = 0;
 
     /* Extract the private key from the PK11Signature */
     if( getPrivateKey(env, this, &privk) != PR_SUCCESS) {
         PR_ASSERT( (*env)->ExceptionOccurred(env) != NULL);
         goto finish;
     }
+    
+    // See if the key is a candidate for PSS signature 
+    if (privk->keyType == rsaKey || privk->keyType == rsaPssKey) {
+        keyTypeOKForPSS = 1;
+    }
+    
+    PRArenaPool *arena = NULL;
+    SECOidTag signingAlg = SEC_OID_UNKNOWN;
 
-    /* Start the signing operation */
-    ctxt = SGN_NewContext(getAlgorithm(env, this), privk);
-	if(ctxt == NULL) {
-		JSS_throwMsg(env, TOKEN_EXCEPTION, "Unable to create signing context");
-		goto finish;
-	}
+    signingAlg = getAlgorithm(env,this);
+
+    if(signingAlg == SEC_OID_PKCS1_RSA_PSS_SIGNATURE && keyTypeOKForPSS) {
```

Comment:
What I'm missing is why are we continuing if the key isn't OK for PSS?

In particular, when `signingAlg == SEC_OID_PKCS1_RSA_PSS_SIGNATURE` and `!keyTypeOKForPSS`, shouldn't we fail?

Reply:
The else branch is just for everything else not PSS. Note that the context is created differently in those two cases.

Reply:
Right, I'm happier with the new check.